packages/adapters/ethers/src/client.ts

```diff
@@ -175,889 +95,394 @@ export class EthersAdapter {
-  } | null = null
-
-  private removeListeners(provider: Provider) {
-    if (this.providerHandlers) {
-      provider.removeListener('disconnect', this.providerHandlers.disconnect)
-      provider.removeListener('accountsChanged', this.providerHandlers.accountsChanged)
-      provider.removeListener('chainChanged', this.providerHandlers.chainChanged)
-      this.providerHandlers = null
+    return {
+      address: accounts[0] as `0x${string}`,
+      chainId: Number(chainId),
+      provider: selectedProvider,
+      type: type as ConnectorType,
+      id
     }
   }
 
-  private setupAuthListeners(authProvider: W3mFrameProvider) {
-    authProvider.onRpcRequest(request => {
-      if (W3mFrameHelpers.checkIfRequestExists(request)) {
-        if (!W3mFrameHelpers.checkIfRequestIsSafe(request)) {
-          this.appKit?.handleUnsafeRPCRequest()
-        }
-      } else {
-        this.handleInvalidAuthRequest()
-      }
-    })
+  public async disconnect(params: AdapterBlueprint.DisconnectParams): Promise<void> {
+    if (!params.provider || !params.providerType) {
+      throw new Error('Provider or providerType not provided')
+    }
 
-    authProvider.onRpcError(() => this.handleAuthRpcError())
-    authProvider.onRpcSuccess((_, request) => this.handleAuthRpcSuccess(_, request))
-    authProvider.onNotConnected(() => this.handleAuthNotConnected())
-    authProvider.onConnect(({ preferredAccountType }) =>
-      this.handleAuthIsConnected(preferredAccountType)
-    )
-    authProvider.onSetPreferredAccount(({ address, type }) => {
-      if (address) {
-        this.handleAuthSetPreferredAccount(address, type)
-      }
-    })
+    switch (params.providerType) {
+      case 'WALLET_CONNECT':
+      case 'AUTH':
+        await params.provider.disconnect()
+        break
+      case 'ANNOUNCED':
+        await this.revokeProviderPermissions(params.provider as Provider)
+        break
+      case 'EXTERNAL':
+        await this.revokeProviderPermissions(params.provider as Provider)
+        break
+      default:
+        throw new Error('Unsupported provider type')
+    }
   }
 
-  private handleInvalidAuthRequest() {
-    this.appKit?.open()
-    setTimeout(() => {
-      this.appKit?.showErrorMessage(W3mFrameRpcConstants.RPC_METHOD_NOT_ALLOWED_UI_MESSAGE)
-    }, 300)
-  }
+  public async getBalance(
+    params: AdapterBlueprint.GetBalanceParams
+  ): Promise<AdapterBlueprint.GetBalanceResult> {
+    const caipNetwork = this.caipNetworks?.find((c: CaipNetwork) => c.id === params.chainId)
 
-  private handleAuthRpcError() {
-    if (this.appKit?.isOpen()) {
-      if (this.appKit?.isTransactionStackEmpty()) {
-        this.appKit?.close()
-      } else {
-        this.appKit?.popTransactionStack(true)
-      }
-    }
-  }
+    if (caipNetwork) {
+      const jsonRpcProvider = new JsonRpcProvider(caipNetwork.rpcUrls.default.http[0], {
+        chainId: caipNetwork.id as number,
+        name: caipNetwork.name
```

Comment:
might be harder to maintain, but wyt about having this an attribute of the adapter? I mean every adapter will have an underlying provider when connected right?